basic_pipeline/8_Mixer.md

```diff
@@ -0,0 +1,232 @@
+```Elixir
+# FILE: lib/elements/Mixer.ex
+
+defmodule Basic.Elements.Mixer do
+ ...
+ defp update_state_and_prepare_actions(state, pads) do
+  {state, buffer_actions} = output_buffers(state)
+  {state, end_of_stream_actions} = send_end_of_stream(state)
+  {state, demand_actions} = demand_on_empty_tracks(state, pads)
+
+  actions = buffer_actions ++ end_of_stream_actions ++ demand_actions
+  {state, actions}
+ end
+ ...
+end
+```
+
+The private function `update_state_and_prepare_actions/2` is defined as described in the steps' list above - it aggregates the actions taken in order to:
++ output the ready buffers
++ send `:end_of_stream` notification if necessary
++ demand on empty tracks
+
+Each of these steps has a corresponding private function.
+```Elixir
+# FILE: lib/elements/Mixer.ex
+
+defmodule Basic.Elements.Mixer do
+ ...
+ defp output_buffers(state) do
+  {buffers, tracks} = prepare_buffers(state.tracks)
+  state = %{state | tracks: tracks}
+  buffer_actions = Enum.map(buffers, fn buffer -> {:buffer, {:output, buffer}} end)
+  {state, buffer_actions}
+ end
+
+ defp prepare_buffers(tracks) do
+  active_tracks =
+    tracks
+    |> Enum.reject(fn {_track_id, track} ->
+      track.status == :finished and track.buffer == nil
+    end)
+    |> Map.new()
+
+  if active_tracks != %{} and
+    Enum.all?(active_tracks, fn {_, track} -> track.buffer != nil end) do
+      {track_id, track} =
+        active_tracks
+    |> Enum.min_by(fn {_track_id, track} -> track.buffer.pts end)
+
+  buffer = track.buffer
+  tracks = Map.put(tracks, track_id, %Track{track | buffer: nil})
+  {buffers, tracks} = prepare_buffers(tracks)
+  {[buffer | buffers], tracks}
+  else
+    {[], tracks}
+  end
+ end
+ ...
+end
+```
+
+In order to output the buffers, we need to fetch the desired buffers - that is what we do with the `prepare_buffers/1` function. Later on, we are simply creating the `:buffer` action, basing on the list of buffers to be output. 
+In the `prepare_bufers/1` we get all the active tracks (by 'active' we mean that there is still an unprocessed buffer in the `Track` structure - independent of the status of that track). If all the active tracks have the buffers we can output the one with the lowest presentation timestamp and recursively call the `prepare_buffers/1` (in case there are some buffers that still need to be output - this can happen in a 'corner case' of processing the buffer from the track in the `:finished` state). Surely, we also need to update the state so that to remove the processed buffers.
+Now let's focus on preparing `:end_of_stream` action:
+```Elixir
+# FILE: lib/elements/Mixer.ex
+
+defmodule Basic.Elements.Mixer do
+ ...
+ defp send_end_of_stream(state) do
+  end_of_stream_actions =
+    if Enum.all?(state.tracks, fn {_, track} -> track.status == :finished end) do
+      [end_of_stream: :output]
+    else
+      []
+    end
+ {state, end_of_stream_actions}
+ end
```

Comment:
hmm... shouldn't the `end_of_stream` be sent when the status is `:finished` and `%Track{}` is empty?

Reply:
As far as I remember we have accepted on the naming convention that the track is in `:finished` state if it has received `:end_of_stream`, no matter if there is still a buffer waiting to be processed. That naming convention is reflected in the way `prepare_buffers` is implemented (there we are not processing only tracks which are `:finished` AND their buffer is equal to nil.